File: maximum-depth-of-binary-tree/radiantchoi.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func maxDepth(_ root: TreeNode?) -> Int {
        var result = 0

        traverse(root, 0, &result)

        return result
    }

    func traverse(_ node: TreeNode?, _ current: Int, _ result: inout Int) {
        guard let node else {
            result = max(current, result)
            return
        }

        let newCurrent = current + 1

        traverse(node.left, newCurrent, &result)
        traverse(node.right, newCurrent, &result)
    }
}

File: maximum-subarray/radiantchoi.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        n = len(nums)

        # 원래라면 dp 배열을 만들어, i번째 원소까지 탐색한 결과 중 최댓값을 각각의 인덱스에 저장
        # 하지만 지금은 최댓값만 남기면 되므로 일차적 최적화
        current = nums[0]
        result = nums[0]

        for i in range(1, len(nums)):
            current = max(current + nums[i], nums[i])
            result = max(result, current)

        return result
